packages/ember/tests/routing/substates_test.js

```diff
@@ -1,1042 +1,880 @@
+      }
+    }));
 
-  templates['foo/bar_loading'] = 'FOOBAR LOADING';
-  templates['foo/bar'] = 'YAY';
+    this.add('route:mom', Route.extend({
+      model() {
+        return {};
+      }
+    }));
 
-  Router.map(function() {
-    this.route('foo', function() {
-      this.route('bar');
-    });
-  });
+    this.add('route:mom.sally', Route.extend({
+      model() {
+        return sallyDeferred.promise;
+      },
 
-  App.ApplicationController = Controller.extend();
+      setupController() {
+      }
+    }));
 
-  let deferred = RSVP.defer();
-  App.FooBarRoute = Route.extend({
-    model() {
-      return deferred.promise;
-    }
-  });
+    this.visit('/grandma/mom/sally');
 
-  bootApplication('/foo/bar');
+    let text = this.$('#app').text();
 
-  equal(jQuery('#app', '#qunit-fixture').text(), 'FOOBAR LOADING', 'foo.bar_loading was entered (as opposed to something like foo/foo/bar_loading)');
+    assert.equal(text, 'LOADING', `
+      The loading template is nested in application template's outlet
+    `);
 
-  run(deferred, 'resolve');
+    this.runTask(() => grandmaDeferred.resolve());
+    text = this.$('#app').text();
 
-  equal(jQuery('#app', '#qunit-fixture').text(), 'YAY');
-});
+    assert.equal(text, 'GRANDMA MOM MOMLOADING', `
+      Mom's child loading route is displayed due to sally's slow promise
+    `);
 
-QUnit.test('Prioritized error substate entry works with preserved-namespace nested routes', function() {
-  expect(2);
+    this.runTask(() => sallyDeferred.resolve());
+    text = this.$('#app').text();
 
-  templates['foo/bar_error'] = 'FOOBAR ERROR: {{model.msg}}';
-  templates['foo/bar'] = 'YAY';
-
-  Router.map(function() {
-    this.route('foo', function() {
-      this.route('bar');
-    });
-  });
+    assert.equal(text, 'GRANDMA MOM SALLY', `
+      Sally template displayed
+    `);
+  }
+  ['@test Enter child loading state of pivot route'](assert) {
+    let deferred = RSVP.defer();
+    this.addTemplate('grandma.loading', 'GMONEYLOADING');
 
-  App.ApplicationController = Controller.extend();
+    this.add('route:mom.sally', Route.extend({
+      setupController() {
+      }
+    }));
 
-  App.FooBarRoute = Route.extend({
-    model() {
-      return RSVP.reject({
-        msg: 'did it broke?'
-      });
-    }
-  });
+    this.add('route:grandma.smells', Route.extend({
+      model() {
+        return deferred.promise;
+      }
+    }));
 
-  throws(function() {
-    bootApplication('/foo/bar');
-  }, function(err) { return err.msg === 'did it broke?'; });
+    return this.visit('/grandma/mom/sally').then(() => {
+      let appController = this.getController('application');
+      assert.equal(appController.get('currentPath'), 'grandma.mom.sally',
+        'Initial route fully loaded'
+      );
 
-  equal(jQuery('#app', '#qunit-fixture').text(), 'FOOBAR ERROR: did it broke?', 'foo.bar_error was entered (as opposed to something like foo/foo/bar_error)');
-});
+      this.visit('/grandma/smells');
```

Comment:
same thing here as above ^, this relies on assertions before and after promises are resolved on `model()` hooks in routes.